src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -1055,17 +1059,66 @@ public DynamicObject classNew(VirtualFrame frame, DynamicObject superclass) {
     @CoreMethod(names = "rb_tr_debug", onSingleton = true, rest = true)
     public abstract static class DebugNode extends CoreMethodArrayArgumentsNode {
 
-        @TruffleBoundary
+        @Child CallDispatchHeadNode toSCall;
+
         @Specialization
-        public Object debug(Object... objects) {
+        public Object debug(VirtualFrame frame, Object... objects) {
             if (objects.length > 1) {
-                System.err.printf("Printing %d values%n", objects.length);
+                printf("Printing %d values%n", objects);
             }
             for (Object object : objects) {
-                System.err.printf("%s @ %s: %s%n", object.getClass(), System.identityHashCode(object), object);
+                final String representation;
+                if (RubyGuards.isRubyString(object)) {
+                    representation = getRubyStringRepresentation((DynamicObject) object);
+                } else if (RubyGuards.isRubyBasicObject(object)) {
+                    representation = getString(object) + " (" +
+                            RopeOperations.decodeRope(StringOperations.rope(callToS(frame, object))) + ")";
+                } else {
+                    representation = getString(object);
+                }
+                printf("%s @ %s: %s%n", object.getClass(), identityHashCode(object), representation);
             }
             return nil();
         }
+
+        private DynamicObject callToS(VirtualFrame frame, Object object) {
+            if (toSCall == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                toSCall = CallDispatchHeadNode.create();
+            }
+
+            return (DynamicObject) toSCall.call(frame, object, "to_s");
+        }
+
+        @TruffleBoundary
+        private int identityHashCode(Object object) {
+            return System.identityHashCode(object);
+        }
+
+        @TruffleBoundary
+        private void printf(String format, Object... objects) {
+            System.err.printf(format, objects.length);
+        }
+
+        @TruffleBoundary
+        private String getString(Object object) {
+            return object.toString();
+        }
+
+        @TruffleBoundary
+        private String getRubyStringRepresentation(DynamicObject object) {
+            final Rope rope = StringOperations.rope(object);
+            final String str = RopeOperations.decodeRope(rope);
```

Comment:
You could use `StringOperations.getString` here.

Reply:
In this case I need the rope as well, to get bytes

Reply:
Indeed. I missed that.